src/chunkserver/scan_manager.cpp

```diff
@@ -21,156 +21,300 @@
     }
-    return;
 }
 
 ScanKey ScanManager::Dequeue() {
     ScanKey key;
-    WriteLockGuard writeGuard(rwLock_);
-    key = *(waitScanSet.begin());
-    waitScanSet.erase(waitScanSet.begin());
+    WriteLockGuard writeGuard(waitSetLock_);
+    key = *(waitScanSet_.begin());
+    waitScanSet_.erase(waitScanSet_.begin());
     return key;
 }
 
-int ScanManager::CancelScanJob(LogicPoolID poolId, CopysetID id) {
+int ScanManager::Run() {
+    scanThread_ = Thread(&ScanManager::Scan, this);
     return 0;
 }
 
-bool ScanManager::IsRepeatReq(LogicPoolID poolId, CopysetID id) {
-    return 0;
+void ScanManager::Scan() {
+    LOG(INFO) << "Starting Scan worker thread.";
+    ScanKey key;
+    while (!toStop_.load(std::memory_order_acquire)) {
+        waitSetLock_.RDLock();
+        if (waitScanSet_.empty()) {
+            waitSetLock_.Unlock();
+            waitInterval_.WaitForNextExcution();
+            continue;
+        }
+        waitSetLock_.Unlock();
+        key = Dequeue();
+        StartScanJob(key);
+        waitInterval_.WaitForNextExcution();
+    }
+    LOG(INFO) << "Scan worker thread stopped.";
 }
 
 void ScanManager::StartScanJob(ScanKey key) {
-    job_.poolId = key.first;
-    job_.id = key.second;
-    job_.type = ScanType::Init;
-    job_.currentChunkId = 0;
-    job_.startTime = 0;
-    GenScanJob();
+    std::shared_ptr<ScanJob> job = std::make_shared<ScanJob>();
+    job->poolId = key.first;
+    job->id = key.second;
+    job->type = ScanType::Init;
+    job->currentChunkId = 0;
+    job->startTime = 0;
+    job->waitingNum = 3;
+    // check copyset exist
+    auto nodePtr = copysetNodeManager_->GetCopysetNode(job->poolId, job->id);
+    if (nullptr == nodePtr) {
+        LOG(ERROR) << "scan copyset failed, copyset node is not found: "
+                   << job->poolId << "," << job->id;
+        return;
+    }
+    dataStore_ = nodePtr->GetDataStore();
+    nodePtr->SetScan(true);
+    jobMapLock_.WRLock();
+    jobs_.emplace(key, job);
+    jobMapLock_.Unlock();
+    GenScanJob(key);
+}
+
+int ScanManager::CancelScanJob(LogicPoolID poolId, CopysetID id) {
+    ScanKey key(poolId, id);
+    // cancel scan job not started
+    {
+        WriteLockGuard writeGuard(waitSetLock_);
+        auto iter = waitScanSet_.find(key);
+        if (iter != waitScanSet_.end()) {
+            waitScanSet_.erase(iter);
+            return 0;
+        }
+    }
+    // cancel scan job started
+    auto job = GetJob(key);
+    if (nullptr != job) {
+        auto nodePtr = copysetNodeManager_->GetCopysetNode(poolId, id);
+        nodePtr->SetScan(false);
+        WriteLockGuard writeGuard(jobMapLock_);
+        jobs_.erase(key);
+    }
+    return 0;
 }
 
-void ScanManager::GenScanJob() {
+void ScanManager::GenScanJob(ScanKey key) {
     bool done = false;
+    auto job = GetJob(key);
+    if (nullptr == job) {
+        return;
+    }
 
     while (!done) {
-        switch (job_.type) {
+        switch (job->type) {
```

Comment:
switch Encapsulation as 
ScanType GenScanJob(ScanKey key)

Reply:
> switch Encapsulation as
> ScanType GenScanJob(ScanKey key)

Now the GenScanJob  mainly contains 'switch'. Why encapsulation another function to do the same thing?